clients/python-wrapper/lakefs/reference.py

```diff
@@ -1,22 +1,179 @@
+from lakefs.object import Object
+from lakefs.object_manager import ObjectManager
+
+
+class Commit(NamedTuple):
+    """
+    NamedTuple representing a lakeFS commit's properties
+    """
+    id: str
+    parents: [str]
+    committer: str
+    message: str
+    creation_date: int
+    meta_range_id: str
+    metadata: Optional[dict[str, str]] = None
+
+
+class Change(NamedTuple):
+    """
+    NamedTuple representing a diff change between to refs in lakeFS
+    """
+    type: Literal["added", "removed", "changed", "conflict", "prefix_changed"]
+    path: str
+    path_type: Literal["common_prefix", "object"]
+    size_bytes: Optional[int]
 
 
 class Reference:
     """
     Class representing a reference in lakeFS.
     """
     _client: Client
-    repo_id: str
-    id: str
+    _repo_id: str
+    _id: str
+    _commit: Optional[Commit] = None
 
     def __init__(self, repo_id: str, ref_id: str, client: Optional[Client] = DefaultClient) -> None:
         self._client = client
-        self.repo_id = repo_id
-        self.id = ref_id
+        self._repo_id = repo_id
+        self._id = ref_id
+
+    @property
+    def repo_id(self) -> str:
+        """
+        Return the repository id for this reference
+        """
+        return self._repo_id
+
+    @property
+    def id(self) -> str:
+        """
+        Returns the reference id
+        """
+        return self._id
+
+    @property
+    def objects(self) -> ObjectManager:
+        """
+        Returns a ObjectManager object for this reference
+        """
+        # TODO: Implement
+
+    @staticmethod
+    def _get_generator(func, *args, max_amount: Optional[int] = None, **kwargs):
+        count = 0
+        has_more = True
+        while has_more:
+            with api_exception_handler():
+                page = func(*args, **kwargs)
+                has_more = page.pagination.has_more
+                for res in page.results:
+                    count += 1
+                    yield res
+                    if max_amount is not None and count >= max_amount:
+                        return
+
+    def log(self, max_amount: Optional[int] = None, **kwargs) -> Generator[lakefs_sdk.Commit]:
+        """
+        Returns a generator of commits starting with this reference id
+        :param max_amount: (Optional) limits the amount of results to return from the server
+        :param kwargs: Additional keyword arguments
+        :raises
+            NotFoundException if reference by this id does not exist
+            NotAuthorizedException if user is not authorized to perform this operation
+            ServerException for any other errors
+        """
+        limit = max_amount is not None
+        return self._get_generator(self._client.sdk_client.refs_api.log_commits,
+                                   self._repo_id, self._id, max_amount=max_amount, limit=limit, **kwargs)
+
+    def _get_commit(self):
+        if self._commit is None:
+            with api_exception_handler():
```

Comment:
Since we use these for basically any method/function that does any I/O, I would suggest using a decorator instead of a context manager. 

Will reduce code complexity by lowering your average indentation :)

Reply:
decorator will require us to wrap the entire function logic under this try catch, will also make adding custom behavior more difficult. 
Generally I like to wrap only the minimally required code with a try-except